homeassistant/components/mqtt.py

```diff
@@ -0,0 +1,170 @@
+        _LOGGER.exception("Error while importing dependency paho-mqtt.")
+        return False
+
+    global MQTT_CLIENT
+
+    MQTT_CLIENT = MQTT(hass,
+                       config[DOMAIN]['broker'],
+                       config[DOMAIN]['port'],
+                       config[DOMAIN]['topic'],
+                       config[DOMAIN]['keepalive'],
+                       config[DOMAIN]['client_id'])
+
+    def stop_mqtt(event):
+        """ Stop MQTT component. """
+        MQTT_CLIENT.stop()
+
+    def start_mqtt(event):
+        """ Launch MQTT component when Home Assistant starts up. """
+        try:
+            MQTT_CLIENT.run()
+
+        except ConnectionRefusedError:
+            _LOGGER.exception("Can't connect to the broker. "
+                              "Please check your settings and the broker"
+                              "itself.")
+            return False
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_mqtt)
+
+    def send_message(call):
+        """ Sending an MQTT message. """
+        subtopic = 'master'
+        complete_topic = 'home-assistant/{}'.format(str(subtopic))
+        MQTT_CLIENT.publish(complete_topic, str(call))
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_mqtt)
+
+    hass.services.register(DOMAIN, MQTT_SEND, send_message)
+    hass.services.register(DOMAIN, EVENT_MQTT_MESSAGE_RECEIVED)
+
+    return True
+
+
+# This is based on one of the paho-mqtt examples:
+# http://git.eclipse.org/c/paho/org.eclipse.paho.mqtt.python.git/tree/examples/sub-class.py
+# pylint: disable=too-many-arguments, invalid-name
+class MQTT(object):
+    """ Implements messaging service for MQTT. """
+    def __init__(self, hass, broker, port, topic, keepalive, clientid=None):
+
+        import paho.mqtt.client as mqtt
+
+        self.hass = hass
+        self._broker = broker
+        self._port = port
+        self._topic = topic
+        self._keepalive = keepalive
+        self.msg = None
+
+        self._mqttc = mqtt.Client(clientid)
+        self._mqttc.on_message = self.mqtt_on_message
+        self._mqttc.on_connect = self.mqtt_on_connect
+        self._mqttc.on_publish = self.mqtt_on_publish
+        self._mqttc.on_subscribe = self.mqtt_on_subscribe
+
+    def mqtt_on_connect(self, mqttc, obj, flags, rc):
+        """ Connect callback """
+        _LOGGER.info('Connected to broker %s', self._broker)
+
+    def mqtt_on_publish(self, mqttc, obj, mid):
+        """ Publish callback """
+        pass
+
+    def mqtt_on_subscribe(self, mqttc, obj, mid, granted_qos):
+        """ Subscribe callback """
+        complete_topic = self._topic + '/#'
+        _LOGGER.info('Subscribed to %s', complete_topic)
+
+    def mqtt_on_message(self, mqttc, obj, msg):
+        """ Message callback """
+        self.msg = '{} {} {}'.format(msg.topic, str(msg.qos), str(msg.payload))
+        self.hass.bus.fire(EVENT_MQTT_MESSAGE_RECEIVED, {
+            'topic': msg.topic,
+            'subtopic': 'test',
```

Comment:
No, was just a test...will be removed. First I was thinking about handling the topic here but it seems to make more sense to do that later in the various platforms.